ember_debug/libs/octane-tree.js

```diff
@@ -0,0 +1,14 @@
+export function makeRenderNodeCloneable(retainObject, renderNode) {
+  let children = renderNode.children.map(child => makeRenderNodeCloneable(retainObject, child));
+
+  return {
+    // TODO add args and bounds
+    value: {
+      isComponent: renderNode.type === 'component',
+      name: renderNode.name,
+      objectId: renderNode.instance ? retainObject(renderNode.instance) : null,
+      viewClass: renderNode.name
+    },
+    children
+  };
+}
```

Comment:
To filter out the outlets:

```ts
function filterNodes(nodes: CapturedRenderNode[]): CapturedRenderNode[] {
  return nodes.map(
}

function filterNode(node: CapturedRenderNode): CapturedRenderNode {
  if (node.type === 'outlet') {
    assert('outlet node should have exactly one child', node.children.length === 1);
    return filterNode(node.children[0]);
  } else {
    return { ...node, children: filterNodes(node.children) };
  }
}
```

Reply:
So this would be filtering out after the fact, not inside the current recursive call?

Reply:
You could work it into the existing one too. Just do the logic inside your make RenderNodeClonable function (and probably rename it).

Reply:
@chancancode that's the issue though, is my brain cannot grok how to do it in the existing `.map` for children.

Reply:
I would think this should work?

```
export function whateverYouCallThisNow(retainObject, renderNode) {
  if (node.type === 'outlet') {
    assert('outlet node should have exactly one child', node.children.length === 1);
    return whateverYouCallThisNow(retainObject, node.children[0]);
  } else {
    let children = renderNode.children.map(child => whateverYouCallThisNow(retainObject, child));

    return {
      // TODO add args and bounds
      value: {
        isComponent: renderNode.type === 'component',
        name: renderNode.name,
        objectId: renderNode.instance ? retainObject(renderNode.instance) : null,
        viewClass: renderNode.name
      },
      children
    };
  }
}
```